src/common/validation.rs

```diff
@@ -123,6 +126,40 @@ impl TransactionValidator {
         Ok(())
     }
 
+    fn validate_program_instructions(&self, message: &Message) -> Result<(), KoraError> {
+        for instruction in &message.instructions {
+            let program_id = message.account_keys[instruction.program_id_index as usize];
+            
+            // Find program config in allowed instructions
+            let program_config = self.allowed_program_instructions
+                .iter()
+                .find(|config| Pubkey::from_str(&config.program_id).unwrap() == program_id)
+                .ok_or_else(|| KoraError::InvalidTransaction(
+                    format!("Program {} not found in allowed instructions", program_id)
+                ))?;
+
+            // If wildcard "*" is present, allow all instructions for this program
+            if program_config.instructions.contains(&"*".to_string()) {
+                continue;
+            }
+
+            // Check if instruction data matches any allowed discriminator
+            let instruction_data = &instruction.data[..8];  // First 8 bytes are the discriminator
+            let instruction_data_vec = instruction_data.to_vec();
+
+            let is_allowed = program_config.discriminators.iter().any(|(disc, _)| {
+                disc == &instruction_data_vec
+            });
```

Comment:
the exact size of the discriminator is only 8 bytes for anchor programs...

Better to check `disc == &instruction.data[..disc.len()]` inside the iterator